File: java/src/test/java/it/giannibombelli/wsc2026/giftcard/application/query/GiftCardQueryServiceTest.java
package it.giannibombelli.wsc2026.giftcard.application.query;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import it.giannibombelli.wsc2026.giftcard.domain.ports.GiftCardRepository;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;

class GiftCardQueryServiceTest {

    @Test
    void shouldReturnGiftCardDetailsWhenGiftCardExists() {
        InMemoryGiftCardRepository repository = new InMemoryGiftCardRepository();
        GiftCardQueryService queryService = new GiftCardQueryService(repository);
        GiftCardId cardId = EntityId.generate(GiftCardId::new);
        Money balance = new Money(new BigDecimal("42.50"));
        GiftCard card = new GiftCard(cardId, balance);
        repository.save(card);

        Optional<GiftCardDetails> result = queryService.findById(cardId);

        assertThat(result).isPresent();
        assertThat(result.get().id()).isEqualTo(cardId.value());
        assertThat(result.get().balance()).isEqualTo(balance);
    }

    @Test
    void shouldReturnEmptyWhenGiftCardDoesNotExist() {
        InMemoryGiftCardRepository repository = new InMemoryGiftCardRepository();
        GiftCardQueryService queryService = new GiftCardQueryService(repository);
        GiftCardId cardId = EntityId.generate(GiftCardId::new);

        Optional<GiftCardDetails> result = queryService.findById(cardId);

        assertThat(result).isEmpty();
    }

    private static final class InMemoryGiftCardRepository implements GiftCardRepository {
        private final Map<UUID, GiftCard> cards = new HashMap<>();

        @Override
        public void save(GiftCard card) {
            cards.put(card.id().value(), card);
        }

        @Override
        public Optional<GiftCard> findById(GiftCardId id) {
            return Optional.ofNullable(cards.get(id.value()));
        }
    }
}

File: typescript/tests/giftcard/application/query/giftCardQueryService.test.ts
import { describe, expect, it } from 'vitest';
import { GiftCardQueryService } from '@/giftcard/application/query/giftCardQueryService.js';
import { GiftCard } from '@/giftcard/domain/giftcard/giftCard.js';
import { GiftCardId } from '@/giftcard/domain/giftcard/giftCardId.js';
import { GiftCardRepository } from '@/giftcard/domain/ports/giftCardRepository.js';
import { Money } from '@/common/domain/primitive/money.js';
import { generateId } from '@/common/domain/identity/entityId.js';

class InMemoryGiftCardRepository implements GiftCardRepository {
    private readonly cards = new Map<string, GiftCard>();

    save(card: GiftCard): void {
        this.cards.set(card.id().value.value, card);
    }

    findById(id: GiftCardId): GiftCard | null {
        return this.cards.get(id.value.value) ?? null;
    }
}

describe('GiftCardQueryService', () => {
    it('should return gift card details when gift card exists', () => {
        const repository = new InMemoryGiftCardRepository();
        const queryService = new GiftCardQueryService(repository);
        const cardId = generateId((value) => new GiftCardId(value));
        const balance = new Money(42.5);
        const card = new GiftCard(cardId, balance);
        repository.save(card);

        const result = queryService.findById(cardId);

        expect(result).not.toBeNull();
        expect(result!.id).toBe(cardId.value.value);
        expect(result!.balance).toBe(balance);
    });

    it('should return null when gift card does not exist', () => {
        const repository = new InMemoryGiftCardRepository();
        const queryService = new GiftCardQueryService(repository);
        const cardId = generateId((value) => new GiftCardId(value));

        const result = queryService.findById(cardId);

        expect(result).toBeNull();
    });
});
